src/compiler/core.ts

```diff
@@ -2590,3 +2590,96 @@ export function isNodeLikeSystem(): boolean {
         && !(process as any).browser
         && typeof require !== "undefined";
 }
+
+/** @internal */
+export interface PrefixSuffixTrie<T extends {}> {
+    iterateAllMatches(input: string): Iterable<T>;
+    set(prefix: string, suffix: string, fn: (value: T | undefined) => T | undefined): void;
+    hasAnyMatch(input: string): boolean;
+}
+
+/** @internal */
+export function createPrefixSuffixTrie<T extends {}>(): PrefixSuffixTrie<T> {
+    interface Trie<T extends {}> {
+        children: Record<string, Trie<T>> | undefined;
+        value: T | undefined;
+    }
+
+    function createTrie<T extends {}>(): Trie<T> {
+        return {
+            children: undefined,
+            value: undefined,
+        };
+    }
+
+    const root = createTrie<Trie<T>>();
+
+    function* iterateAllMatches(input: string) {
+        let node = root;
+
+        if (node.value) {
+            yield* iterateSuffix(node.value, 0);
+        }
+
+        for (let i = 0; i < input.length; i++) {
+            const child = node.children?.[input[i]];
+            if (!child) break;
+            if (child.value) {
+                yield* iterateSuffix(child.value, i + 1);
+            }
+            node = child;
+        }
+
+        return;
+
+        function* iterateSuffix(node: Trie<T>, start: number) {
+            if (node.value) {
+                yield node.value;
+            }
+
+            for (let i = input.length - 1; i >= start; i--) {
+                const child = node.children?.[input[i]];
+                if (!child) break;
+                if (child.value) {
+                    yield child.value;
+                }
+                node = child;
+            }
+        }
+    }
+
+    function set(prefix: string, suffix: string, fn: (value: T | undefined) => T | undefined) {

```

Comment:
An interesting thing about building `trie`s dynamically is that they can grow quite large in memory consumed rather quickly when you do something like this and always create new `trie` nodes for every new match state. When you're matching a whole dictionary, for example, this can quickly become untenably large. All is not lost, however - there are a large number of sub-`trie`s that are identical (for example, the `-es` suffix trie), and you can save quite a bit of space at the cost of some time by compressing the trie by unifying all those subtrees. So rather than having a `trie` like
```mermaid
stateDiagram-v2
    1:_ 
    2:_
    3:_
    4:_
    5:_
    6:_
    7:_
    8:_
    9:_
    10:_
    11:_
    12:_
    [*] --> 1 : s
    1 --> 2 : p
    2 --> 3 : a
    3 --> 4 : c
    4 --> 5 : e
    5 --> 6 : s
    6 --> [*] : End of string
    [*] --> 7 : p
    7 --> 8 : l
    8 --> 9 : a
    9 --> 10 : c
    10 --> 11 : e
    11 --> 12 : s
    12 --> [*] : End of string
```
you compress it like so
```mermaid
stateDiagram-v2
    s:_
    p:_
    a:_
    c:_
    e:_
    s`:_
    p`:_
    l:_
    [*] --> s : s
    s --> p : p
    p --> a : a
    a --> c : c
    c --> e : e
    e --> s` : s
    s` --> [*] : End of string
    [*] --> p` : p
    p` --> l : l
    l --> a : a
```
or, better yet,
```mermaid
stateDiagram-v2
    s:_
    p:_
    p`:_
    l:_
    [*] --> s : s
    s --> p : p
    p --> [*] : aces
    [*] --> p` : p
    p` --> l : l
    l --> [*] : aces
```
Which, a quick search seems to indicate is referred to either as a "compressed trie" or just a "suffix tree". Extra memory efficiency, since JS is GC'd, can change the breakpoint where it's worth it to swap between the naive scan and the trie-based structure.

Reply:
Yeah, I thought about compressing the trie into a proper radix tree, but wasn't sure if that would be worth it in the end (and the implementation does get more complicated). But I can give it a try.

Reply:
The main problem is just that the edges being strings means that I don't know how to walk down them without iterating over each of them...

Reply:
I guess radix tress ensure that the children are keyed on equal length prefixes, so that would work.

However, your example coming together like this I don't think works out for our use case here; otherwise as we walk down the graph, we'll merge together with template literals that don't actually match (because they came from another branch) and not reduce the number of inferences that much.

A proper radix tree doesn't converge like this so would at least still be optimal, I think, so I may attempt that if I have a change.